Added grid generation utlity for ConfigSpaces

RaghuSpaceRajan opened the pull request.

d060030 commented:
Two minor remarks:
- if the defined, the grid generation could utilize the quantization for UniformFloatHyperparameter and UniformIntegerHyperparameter instead of relying on num_steps_dict
- the function's description could mention that constraints in the config space are ignored when generating the grid

RaghuSpaceRajan commented:
Thanks for the feedback!

- I did not know about the quantization factor. Would be good to have it as a default step size for the grid.
- I updated the documentation to reflect that conditional HyperParameters are ignored currently. I wanted to add those, but probably in the future at some point.

Neeratyoy commented:
Hi, is this going to be merged?

RaghuSpaceRajan commented:
> We should probably have a look at whether that's possible. @RaghuSpaceRajan is there a reason you chose this to be a class and not a function?

On looking at it again, I think it was because there was a function called `get_grid_point` (in my local workspace) to get a **specific point** in the generated grid using a list of indices for the different HPs using their "natural" order in `ConfigSpace`, e.g. passing `[0, 2, 1]` for a grid of `3` HPs would return the grid point containing the 1st HP's 0th grid value, the 2nd HP's 2nd grid value and the 3rd HP's 1st grid value (_0-indexed_). I was storing the generated grid as a member variable to not repeat grid creation every time I wanted to query a particular grid point. But then I added grid generation for conditional spaces and "counting" in the generated grid becomes much more involved and I removed `get_grid_point`.

If you prefer, I can make grid generation be a function again, @mfeurer.

> Also, to have this merged we'd need some unit tests, especially for testing the behavior when a hyperparameter has a given quantization.

I will add these soon.

mfeurer commented on ConfigSpace/util.pyx:
```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
                 raise TypeError("Unknown hyperparameter type %s" % type(param))
     return configuration
+
+cimport cython
```
Could you please put imports at the top of the file?

RaghuSpaceRajan replied:
Done

mfeurer commented on ConfigSpace/util.pyx:
```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
     return configuration
+
+cimport cython
+@cython.boundscheck(True)  # Activate bounds checking
```
Great that you figured out how to only change the settings for only this function!

RaghuSpaceRajan replied:
Yes, the link here: [https://cython.readthedocs.io/en/latest/src/userguide/numpy_tutorial.html#tuning-indexing-further](https://cython.readthedocs.io/en/latest/src/userguide/numpy_tutorial.html#tuning-indexing-further) seems to suggest array lookups _are_ slowed down by deactivating those.

mfeurer commented on ConfigSpace/util.pyx:
```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
+@cython.boundscheck(True)  # Activate bounds checking
+@cython.wraparound(True)  # Activate negative indexing
+def generate_grid(configuration_space: ConfigurationSpace,
+                 num_steps_dict: Union[None,  Dict[str, int]] = None,
```
In case you want to simplify this you can use `Optional[Dict[str, int]]`.

RaghuSpaceRajan replied:
It still seems to require specifying the default of `None` though.

mfeurer commented on ConfigSpace/util.pyx:
```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
+        List containing Configurations. It is a cartesian product of tuples of HyperParameter values. Each tuple lists the possible values taken by the corresponding HyperParameter. Within the cartesian product, in each element, the ordering of HyperParameters is the same for the OrderedDict within the ConfigurationSpace.
+    """
+
+    def get_value_set(num_steps_dict, hp_name):
```
Could you please add type annotations here, too?

RaghuSpaceRajan replied:
Done

mfeurer commented on ConfigSpace/util.pyx:
```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
+                grid_points = np.exp(grid_points)
+
+            # Avoiding rounding off issues
+            print(grid_points)
```
Could you please remove this print call?

RaghuSpaceRajan replied:
Done

mfeurer commented on ConfigSpace/util.pyx:
```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
+        '''
+        Gets values along the grid for a particular hyperparameter.
+
+        Uses the num_steps_dict to determine number of grid values for UniformFloatHyperparameter and UniformIntegerHyperparameter. If these values are not present in num_steps_dict, the quantization, q, objects of these classes will be used to divide the grid. NOTE: If q is None, this results in np.arange using None for its step size parameter which results in a quantization of 1.
```
I'm wondering whether it would make more sense to fail if neither the number of steps nor q is given instead of assuming the user meant to set the value to 1. What do you think?

RaghuSpaceRajan replied:
Yes, I had thought about throwing a warning, especially if the number of possible grid points is "too many". But probably, an Error is better. Added that now.

mfeurer commented on ConfigSpace/util.pyx:
```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
+        else:
+            raise TypeError("Unknown hyperparameter type %s" % type(param))
+
+    def get_cartesian_product(value_sets, hp_names):
```
Same as above, could you please add type annotations here?

RaghuSpaceRajan replied:
Done

mfeurer commented on ConfigSpace/util.pyx:
```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
+        '''
+        grid = []
+        import itertools
+        for i, element in enumerate(itertools.product(*value_sets)):
```
Do you need the `i` for something? If not, could you please drop the `enumerate` here?

RaghuSpaceRajan replied:
Done

mfeurer commented on ConfigSpace/util.pyx:
```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
+        try:
+            grid_point = Configuration(configuration_space, unchecked_grid_pts[0])
+            checked_grid_pts.append(grid_point)
+        except ValueError as e:
```
I'm wondering whether this check is too broad and whether we should add new exceptions for this case?

RaghuSpaceRajan replied:
I could check for the error message in `ValueError` to match `Active hyperparameter '%s' not specified!` and print out an info/warning message with the `ValueError` that was caught so that it is not missed. And when the text doesn't match we could rethrow the error.

mfeurer replied:
I think raising if the text doesn't match is sufficient. No need to print a warning/info if it matches.

RaghuSpaceRajan replied:
Done.

mfeurer commented on ConfigSpace/util.pyx:
```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
+        hp_names.append(hp_name)
+
+    # Create a Cartesian product of above allowed values for the HPs. Hold them in an "unchecked" list because some of the conditionally dependent HPs may become active for some of the elements of the Cartesian product and in these cases creating a Configuration would throw an Error (see below).
+    unchecked_grid_pts = get_cartesian_product(value_sets, hp_names) # Creates a list of Configuration dicts
```
Could you please use a double ended queue (`collections.deque`) here to make the code a bit more intuitive?

RaghuSpaceRajan replied:
Done

mfeurer commented on ConfigSpace/util.pyx:
```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
+            for hp_name in new_active_hp_names:
+                value_sets.append(get_value_set(num_steps_dict, hp_name))
+                hp_names.append(hp_name)
+            if len(new_active_hp_names) > 0: # this check might not be needed, as there is always going to be a new active HP when in this except block? #TODO
```
you could add an else that should never be triggered just to be sure?

RaghuSpaceRajan replied:
And throw an Error inside that else?

mfeurer replied:
Yes, exactly.

RaghuSpaceRajan replied:
Done.

mfeurer commented on ConfigSpace/util.pyx:
```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
+    # Create a Cartesian product of above allowed values for the HPs. Hold them in an "unchecked" list because some of the conditionally dependent HPs may become active for some of the elements of the Cartesian product and in these cases creating a Configuration would throw an Error (see below).
+    unchecked_grid_pts = get_cartesian_product(value_sets, hp_names) # Creates a list of Configuration dicts
+    checked_grid_pts = []
+    conditional_grid_lens = []
```
Could you briefly explain what this variable is used for?

RaghuSpaceRajan replied:
It was more for debug purposes, to check the lengths of "sub-grids" generated in different parts of a conditional `ConfigurationSpace`. I have removed it now.

mfeurer commented on test/test_util.py:
```diff
@@ -342,3 +342,74 @@ def test_fix_types(self):
+
+        # Check randomly pre-selected values in the generated_grid
+        # 2 * 1 * 11 * 6 * 3 total diff. possible configurations
+        self.assertEqual(len(generated_grid), 396)
```
Hey, how about you check a few configurations, such as the first and the last one?

RaghuSpaceRajan replied:
Done

mfeurer replied:
Great, could you do me another favor and compare the whole dictionary at once? That would make the code more readable.

RaghuSpaceRajan replied:
Done.

mfeurer commented on test/test_util.py:
```diff
@@ -342,3 +342,74 @@ def test_fix_types(self):
+        self.assertEqual(generated_grid[45].get_dictionary()['const1'], 4)
+        # The 2 most frequently changing HPs (int1 and ord1) have 3 * 6 = 18 different values for
+        # each value of float1, so the 4th value of float1 of -0.4 is reached after
+        # 3 * 18 = 54 values in the generated_grid (and remains the same for the next 18 values):
```
If it remains the same for the next 18 values, could you maybe check that?

RaghuSpaceRajan replied:
Done

mfeurer commented on test/test_util.py:
```diff
@@ -342,3 +342,74 @@ def test_fix_types(self):
+        int1 = UniformIntegerHyperparameter(name='int1', lower=10, upper=100, log=True)
+        ord1 = OrdinalHyperparameter(name='ord1', sequence=['1', '2', '3'])
+
+        cs.add_hyperparameters([float1, int1, cat1, ord1, const1])
```
Could you please check for a few extreme cases, too? Such as only numerical, only categorical, only constant, only a single hyperparameter, no hyperparameters yet?

RaghuSpaceRajan replied:
For the case of no hyperparameters, in `get_cartesian_product`, `itertools.product()` returns a single empty tuple element (I expected 0 elements) here: [https://github.com/RaghuSpaceRajan/ConfigSpace/blob/19144c34b4b0cce2e9df1bb66943bbd17392c818/ConfigSpace/util.pyx#L566]() which leads to a single empty Configuration being generated in the grid. Should we check for this and make sure no configuration is returned?

mfeurer replied:
> Should we check for this and make sure no configuration is returned?

Yes, I think so.

mfeurer commented on test/test_util.py:
```diff
@@ -342,3 +342,74 @@ def test_fix_types(self):
+        cs2.add_hyperparameters([float2_cond])
+        cond_3 = GreaterThanCondition(float2_cond, int2_cond, 50)
+        cs2.add_conditions([cond_3])
+        # print(cs2)
```
Could you please drop the commented print statements?

RaghuSpaceRajan replied:
Done

mfeurer commented on test/test_util.py:
```diff
@@ -342,3 +342,74 @@ def test_fix_types(self):
+                                LessThanCondition(cat1_cond, float1, 0.5)
+                                )
+        cs2.add_conditions([cond_2])
+        float2_cond = UniformFloatHyperparameter(name='float2_cond',
```
Could you please add a comment that this is a 2nd-level dependency so that it's immediately clear that this is tested?

RaghuSpaceRajan replied:
It was mentioned above in the code. I mentioned it again here.

mfeurer commented on test/test_util.py:
```diff
@@ -342,3 +342,74 @@ def test_fix_types(self):
+        # RR: I manually generated the grid and verified the values were correct.
+        # Here, we test that a few randomly chosen values in the generated grid
+        # correspond to the ones I checked.
+        self.assertEqual(generated_grid[3].get_dictionary()['int1'], 1000)
```
As above, could you please check maybe the first and last configuration completely?

RaghuSpaceRajan replied:
Done
